components/builder-depot/src/server.rs

```diff
@@ -2064,6 +2113,186 @@ pub fn download_latest_builder_key(req: &mut Request) -> IronResult<Response> {
+            db_priv_request.set_origin(origin.take_name());
+        }
+        Err(err) => return Ok(render_net_error(&err)),
+    }
+    let priv_key = match route_message::<
+        OriginPrivateEncryptionKeyGet,
+        OriginPrivateEncryptionKey,
+    >(req, &db_priv_request) {
+        Ok(key) => {
+            let key_str = from_utf8(key.get_body()).unwrap();
+            match BoxKeyPair::secret_key_from_str(key_str) {
+                Ok(key) => key,
+                Err(e) => {
+                    return Ok(Response::with(
+                        (status::UnprocessableEntity, format!("{}", e)),
+                    ))
+                }
+            }
+        }
+        Err(err) => return Ok(render_net_error(&err)),
+    };
+
+    let (name, rev) = match parse_name_with_rev(secret_metadata.sender) {
+        Ok(val) => val,
+        Err(e) => {
+            return Ok(Response::with((
+                status::UnprocessableEntity,
+                format!("Failed to parse name and revision: {}", e),
+            )))
+        }
+    };
+
+    debug!("Using key {:?}-{:?}", name, &rev);
+
+    // fetch the public origin encryption key from the database
+    let mut db_pub_request = OriginPublicEncryptionKeyGet::new();
+    match helpers::get_origin(req, origin) {
+        Ok(mut origin) => {
+            db_pub_request.set_owner_id(origin.get_owner_id());
+            db_pub_request.set_origin(origin.take_name());
+            db_pub_request.set_revision(rev.clone());
+        }
+        Err(err) => return Ok(render_net_error(&err)),
+    }
+    let pub_key =
+        match route_message::<OriginPublicEncryptionKeyGet, OriginPublicEncryptionKey>(
+            req,
+            &db_pub_request,
+        ) {
+            Ok(key) => {
+                let key_str = from_utf8(key.get_body()).unwrap();
+                match BoxKeyPair::public_key_from_str(key_str) {
+                    Ok(key) => key,
+                    Err(e) => {
+                        return Ok(Response::with(
+                            (status::UnprocessableEntity, format!("{}", e)),
+                        ))
+                    }
+                }
+            }
+            Err(err) => return Ok(render_net_error(&err)),
+        };
+
+    let box_key_pair = BoxKeyPair::new(name, rev.clone(), Some(pub_key), Some(priv_key));
+
+    debug!("Decrypting string: {:?}", &secret_metadata.ciphertext);
+
+    // verify we can decrypt the message
+    match box_key_pair.decrypt(&secret_metadata.ciphertext, None, None) {
+        Ok(_) => (),
+        Err(e) => {
+            return Ok(Response::with(
+                (status::UnprocessableEntity, format!("{}", e)),
+            ))
+        }
+    };
+
+    let mut secret = OriginSecret::new();
+    secret.set_name(body.name);
+    secret.set_value(body.value.clone());
+    match helpers::get_origin(req, &origin) {
+        Ok(origin) => secret.set_origin_id(origin.get_id()),
```

Comment:
Some comment here re: caching values from the first call to `helpers::get_origin`.